src/msca/metrics/main.py

```diff
@@ -0,0 +1,475 @@
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+            case ErrorMetric.MEDIAN_ABSOLUTE_ERROR:
+                return metrics.median_absolute_error(
+                    y_true=obs_values,
+                    y_pred=pred_values,
+                    sample_weight=weight_values
+                )
+            case _:
+                raise ValueError(f"Unsupported metric type: {self}")
+    
+    def _eval_grouped(
+        self, 
+        data: pd.DataFrame, 
+        obs: str, 
+        pred: str, 
+        weights: str, 
+        groupby: List[str]
+    ) -> pd.DataFrame:
+        """
+        Calculate error metrics for each group in the DataFrame.
+        
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Input DataFrame
+        obs : str
+            Observed values column name
+        pred : str
+            Predicted values column name
+        weights : str
+            Weights column name
+        groupby : List[str]
+            Grouping column names
+            
+        Returns
+        -------
+        pd.DataFrame
+            DataFrame with groupby columns and calculated metric column
+        """
+        df = data.copy()
+        metric_column_name = f"{self.value}_metric"
+        
+        def calculate_group_metric(group):
+            """Calculate metric for a single group."""
+            try:
+                # Check if required columns exist in the group
+                required_columns = [obs, pred, weights]
+                missing_columns = [col for col in required_columns if col not in group.columns]
+                if missing_columns:
+                    print(f"Missing columns in group: {missing_columns}")
+                    print(f"Available columns: {group.columns.tolist()}")
+                    return np.nan
+                
+                # Extract arrays directly from the group DataFrame
+                obs_values = group[obs].values
+                pred_values = group[pred].values
+                weight_values = group[weights].values
+                
+                # Calculate metric using the single metric calculation method
+                return self._calculate_single_metric(obs_values, pred_values, weight_values)
+                        
+            except Exception as e:
+                print(f"Error in group calculation: {e}")
+                print(f"Group shape: {group.shape if hasattr(group, 'shape') else 'No shape'}")
+                if hasattr(group, 'columns'):
+                    print(f"Group columns: {group.columns.tolist()}")
+                return np.nan
+        
+        # Group by specified columns and calculate metrics
+        grouped_results = df.groupby(groupby).apply(calculate_group_metric, include_groups=False).reset_index()
+        grouped_results.columns = list(groupby) + [metric_column_name]
```

Comment:
Let's please use rename, this is error prune if the order of the columns somehow messed up.